task_processor/management/commands/runprocessor.py

```diff
@@ -0,0 +1,110 @@
+import logging
+import signal
+import time
+import typing
+from argparse import ArgumentParser
+from datetime import timedelta
+
+from django.core.management import BaseCommand
+from django.utils import timezone
+
+from task_processor.task_registry import registered_tasks
+from task_processor.thread_monitoring import (
+    clear_unhealthy_threads,
+    write_unhealthy_threads,
+)
+from task_processor.threads import TaskRunner
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        signal.signal(signal.SIGINT, self._exit_gracefully)
+        signal.signal(signal.SIGTERM, self._exit_gracefully)
+
+        self._threads: typing.List[TaskRunner] = []
+        self._monitor_threads = True
+
+    def add_arguments(self, parser: ArgumentParser):
+        parser.add_argument(
+            "--numthreads",
+            type=int,
+            help="Number of worker threads to run.",
+            default=5,
+        )
+        parser.add_argument(
+            "--sleepintervalms",
+            type=int,
+            help="Number of millis each worker waits before checking for new tasks",
+            default=2000,
+        )
+        parser.add_argument(
+            "--graceperiodms",
+            type=int,
+            help="Number of millis before running task is considered 'stuck'.",
+            default=20000,
+        )
+        parser.add_argument(
+            "--queuepopsize",
+            type=int,
+            help="Number of tasks each worker will pop from the queue on each cycle.",
+            default=10,
+        )
+
+    def handle(self, *args, **options):
+        num_threads = options["numthreads"]
+        sleep_interval_ms = options["sleepintervalms"]
+        grace_period_ms = options["graceperiodms"]
+        queue_pop_size = options["queuepopsize"]
+
+        self._threads.extend(
+            [
+                TaskRunner(
+                    sleep_interval_millis=sleep_interval_ms,
+                    queue_pop_size=queue_pop_size,
+                )
+                for _ in range(num_threads)
+            ]
+        )
+
+        logger.info(
+            "Processor starting. Registered tasks are: %s",
+            list(registered_tasks.keys()),
+        )
+
+        for thread in self._threads:
+            thread.start()
+
+        clear_unhealthy_threads()
+        while self._monitor_threads:
+            time.sleep(1)
+            unhealthy_threads = self._get_unhealthy_threads(
+                ms_before_unhealthy=grace_period_ms + sleep_interval_ms
+            )
+            if unhealthy_threads:
+                write_unhealthy_threads(unhealthy_threads)
+
+        [t.join() for t in self._threads]
```

Comment:
At first glance it doesn't seem possible to exit this while loop and enter the joining code below. Can you add a comment explaining what's going on?

Reply:
Hmm... that's interesting because I'd rather the code be self documenting where possible and a quick scan of the class for the `self._monitor_threads` attribute shows me that it's set to `False` in the `_exit_gracefully` method, which (as can be seen in `__init__`) is hooked into the os termination signals. 

I'd really rather not add a comment here as it is liable to become stale but, since you weren't clear, I'm interested if there's something else we can do to clarify things.

Reply:
Ahhh I see. This only exits on `signal.SIGINT` or `signal.SIGTERM` and you're doing a graceful shutdown with the joining code at the end. 

Honestly, now that I understand it I understand why you don't want to comment on it, but in reality adding a comment at the top of the while loop that says something like "Loop until receiving SIGINT or SIGTERM signal" would make things easier. But feel free to disregard this recommendation. I don't have a better code suggestion than what you have there.

Reply:
I'm going to choose to ignore this for now as I think it could be at risk of ending up stale if there are other things in the future that set the `_monitor_threads` attribute.